### PR TITLE
Skip symlink tests when symlinks cannot be created

### PR DESCRIPTION
**Pull request by ladipro**
### Context

Tests currently don't pass on a clean Windows machine in a non-admin command prompt. Even on Win11 with developer mode enabled (and on a Microsoft DevBox VM, no less), the OS refuses to create symlinks unless the process runs elevated.

### Changes Made

Made the affected tests detect the ability to create symlinks and skip if the `kernel32!CreateSymbolicLink` call is failing.

### Testing

Unit tests in admin and non-admin command prompts.

### Notes

As much as I would like to have maximum possible coverage by running the tests unconditionally, I believe that a good developer experience is more important here. Specifically I do not want to require contributors to run elevated or jump through hoops to make the OS cooperate. This is a niche functionality anyway.

### CODE REVIEW

**Comment by ladipro** [comment]
Related to #8212 although I don't see `DoNotFollowRecursiveSymlinks` failing on my machine.

**Comment by vlada-shubina** [comment]
Just hit same issue with `BinaryLoggerShouldEmbedSymlinkFilesViaTaskOutput`

**Review by vlada-shubina — COMMENTED** [review]
Would it be better to create restrictive custom `Fact` - `Theory` implementation?

Example: [`WindowsOnlyTheory`](https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Tests/Microsoft.NET.TestFramework/Attributes/WindowsOnlyTheoryAttribute.cs)

File: src/Tasks.UnitTests/Copy_Tests.cs [hunk 1]
```diff
@@ -2464,5 +2472,22 @@ internal override void ErrorIfLinkFailedCheck()
...
+
+            string symlinkFile = FileUtilities.GetTemporaryFile();
+            string errorMessage = null;
+            if (NativeMethodsShared.MakeSymbolicLink(symlinkFile, sourceFile, ref errorMessage))
+            {
+                File.Delete(symlinkFile);
+                return true;
+            }
+            return false;
+        }
```
Comment (vlada-shubina) [inline src/Tasks.UnitTests/Copy_Tests.cs:2491]:
```suggestion
        private bool IsSymlinkingSupported(string sourceFile)
        {
            if (!NativeMethodsShared.IsWindows)
            {
                return true;
            }

            string symlinkFile = FileUtilities.GetTemporaryFile();
            string errorMessage = null;
            if (NativeMethodsShared.MakeSymbolicLink(symlinkFile, sourceFile, ref errorMessage))
            {
                File.Delete(symlinkFile);
                return true;
            }
            return false;
        }
```

Can it be the case when SymbolicLink can be created for one file but not for another file?
The solution with custom fact attribute may be as:
```
    public class SupportsSymbolicLinksFactAttribute : TheoryAttribute
    {
        public SupportsSymbolicLinksFactAttribute ()
        {
           if (!NativeMethodsShared.IsWindows)
            {
                return true;
            }

            string symlinkFile = FileUtilities.GetTemporaryFile();
            string sourceFile = FileUtilities.GetTemporaryFile();
            string errorMessage = null;
            if (NativeMethodsShared.MakeSymbolicLink(symlinkFile, sourceFile, ref errorMessage))
            {
                File.Delete(symlinkFile);
                return;
            }
            this.Skip = "This test requires support of symbolic link to be run.";
        }
    }

```

Comment (ladipro) [reply src/Tasks.UnitTests/Copy_Tests.cs:2491]: Yes, this would be much cleaner. Let me try it..

Comment (ladipro) [reply src/Tasks.UnitTests/Copy_Tests.cs]: Seems to be working. Thank you!

File: src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            string sourceFile = FileUtilities.GetTemporaryFile();
+            string destinationFile = FileUtilities.GetTemporaryFile();
+            try
+            {
+                File.Create(sourceFile).Dispose();
+
+                string? errorMessage = null;
+                if (!NativeMethodsShared.MakeSymbolicLink(destinationFile, sourceFile, ref errorMessage))
+                {
+                    Skip = "This test requires symbolic link support to run.";
```
Comment (JanKrivanek) [inline src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs:35]: I'd make this more actionable. Something like "Missing rights to create symbolic links. Need to be run elevated or under development mode (https://learn.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development)"

**Review by JanKrivanek — APPROVED** [review]
Looks great!

I'd possibly consider rewording of the message, but overal it's far better experience for new devs

**Comment by rainersigwald** [comment]
> Even on Win11 with developer mode enabled (and on a Microsoft DevBox VM, no less)

Are you sure this was the case? On my devbox developer mode was off and turning it on fixed symlink creation.

I know @Forgind had strong opinions about this.

**Comment by ladipro** [comment]
image

And I rebooted after that just in case. Is this the developer mode you have enabled or am I off?

**Comment by rainersigwald** [comment]
Weird, yeah, that was it. I didn't have to reboot:

image

**Review by rainersigwald — APPROVED** [review]
LGTM, though if we have the ability to detect that we're in an official CI context I'd love it if these would be enabled and tell us if symlinks are disabled in that environment (so we don't lose test coverage entirely).

**Comment by ladipro** [comment]
Aha! We're looking at an actual code bug 🤦 

https://github.com/dotnet/msbuild/blob/5a635cf1218a5d7836114de1e392e7613d569675/src/Framework/NativeMethods.cs#L1051

Win11 does not identify itself as Major == 11 on desktop. This changes things a bit. I'll bring this PR to Monday's review call.

**Comment by Forgind** [comment]
> Aha! We're looking at an actual code bug 🤦
> 
> https://github.com/dotnet/msbuild/blob/5a635cf1218a5d7836114de1e392e7613d569675/src/Framework/NativeMethods.cs#L1051
> 
> Win11 does not identify itself as Major == 11 on desktop. This changes things a bit. I'll bring this PR to Monday's review call.

I copied that from:
https://github.com/dotnet/runtime/blob/861a886730cd0925ea4e75084d80c2730857b1ff/src/libraries/Common/src/Interop/Windows/Kernel32/Interop.CreateSymbolicLink.cs#L43

As I understand it, Win11 initially shipped as 10.0.22000, which means it should be caught by the second case. The first case was to ensure future windows versions are also included.

Was the second case not firing for you? If so, we may need to reach out to runtime folks. I definitely agree we want users to be able to build MSBuild and run tests and see them succeed without doing anything special with their machines, but I also want good test coverage, and I'm a bit uncomfortable with skipping those tests, primarily because no one looks at skipped tests. We were essentially skipping them before, and we didn't realize we had a serious product bug for years.

**Comment by ladipro** [comment]
I misspoke, Windows pretends to be `6.2.9200.0` when running these tests. So nothing to do with Win11 (as you wrote, the Build check should catch it), the relevant test executable just needs to mark itself as "Win10+ compatible" for this condition to work. I am planning to work on a fix on Monday.

**Comment by Forgind** [comment]
> I misspoke, Windows pretends to be `6.2.9200.0` when running these tests. So nothing to do with Win11 (as you wrote, the Build check should catch it), the relevant test executable just needs to mark itself as "Win10+ compatible" for this condition to work. I am planning to work on a fix on Monday.

That makes much more sense, thanks for clarifying. And that sounds like a PR that would resolve the problem and make me feel much more comfortable 🙂

**Comment by ladipro** [comment]
With the fix we still won't have tests that work out of the box, which I think is a downer. We should document it at the very least, analogous to how we ask developers to enable Windows long path support.

**Comment by Forgind** [comment]
That makes me think I should get back to the system state test pr. That would ideally help with messaging around why tests are failing.

AR-May brought up the point that that only helps if the first test you look at is the new test. I don't think this is possible, but it would be amazing if we could have SystemState_Tests that validate that your computer is set up the way we want it to be, and each has a specific set of tests associated with it such that if the SystemState_Test passes, those tests should pass. If the SystemState_Test fails, it fails noisily but with a very clear error, then automatically causes you to skip any dependent tests.

My original idea was that if you're having trouble figuring out a test, you might skip past it and look at the SystemState_Test, then realize what was really wrong. That's a lot easier but also a worse UI. It seems like the general problem of "I want to build MSBuild and run tests on a fresh machine" has gotten a lot more attention lately, so maybe it could be worth investing more in something like the skip-if-test-fails idea.

**Comment by JanKrivanek** [comment]
> With the fix we still won't have tests that work out of the box, which I think is a downer. We should document it at the very least, analogous to how we ask developers to enable Windows long path support.

documented here:
https://github.com/dotnet/msbuild/blob/main/documentation/wiki/Building-Testing-and-Debugging-on-Full-Framework-MSBuild.md?plain=1#L28-L30

**Review by Forgind — APPROVED** [review]
Sorry; never sent this.

File: src/Build.UnitTests/BackEnd/DebugUtils_tests.cs [hunk 3]
```diff
@@ -17,21 +17,22 @@ public class DebugUtils_Tests
         [Fact]
         public void DumpExceptionToFileShouldWriteInTempPathByDefault()
         {
-            Directory.GetFiles(Path.GetTempPath(), "MSBuild_*failure.txt").ShouldBeEmpty();
```
[the remark this answers is not in the document]
Comment (ladipro) [reply src/Build.UnitTests/BackEnd/DebugUtils_tests.cs:20]: I will change the name, though, `DumpExceptionToFileShouldWriteInTempPathByDefault` does not match the implementation anymore.

File: src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs (hunk 2 above)
Comment (ladipro) [reply src/UnitTests.Shared/RequiresSymbolicLinksFactAttribute.cs]: Done, thank you!

**Comment by Forgind** [comment]
This has conflicts now and also will conflict with #8378

**Review by rokonec — APPROVED** [review]
Nice

**Comment by ladipro** [comment]
> This has conflicts now and also will conflict with #8378
> 
> I think it's otherwise good to go, though; is that true?

I have resolved the conflicts. This should be good to merge. Thank you!

**Pull request merged**